### DISCUSSION TITLE
SDK onFailedRefresh

### DISCUSSION THREAD

**Discussion by NilsBaumgartner1994** [directus/directus#21258] [Feature Requests]
### Summary

Catching when a logged in user is no longer authenticated or when the refresh_token is not valid anymore.

### Motivation

When the user is not authenticated anymore we should redirect him to the login page

### Detailed Design

Provide:
- a better documentation if this exists
- Or provide a small callback hook in Directus client.
- 

### Must Have:
- a possibility to catch if the user of offline

### Drawbacks

-

### Alternatives

-

### Adoption Strategy

Usage of transport wrapper as before?

**Comment by ched-dev** [comment]
I think you need to fill this out more. It's unclear what you are referring to. Do you mean in the SDK or data studio app?

**Reply by NilsBaumgartner1994** [reply]
The SDK

**Reply by ched-dev** [reply]
What login page would you redirect them back to? If you're using the SDK we don't know what login page to send to. You could have implemented your own login page. If we redirect them back to the Directus login page it could be inappropriate for some cases. For example, I have a front end using the SDK with my own login and no Directus app access, so I wouldn't want it to be redirected to Directus login page because my users should never see the admin panel.

I think this is something each implementation would handle themselves ideally.

**Reply by NilsBaumgartner1994** [reply]
Exactly that is the problem :-) but I would like to have something like this:

```
const logoutCallback = () => {}
const client = new createDirectus({onLogout: logoutCallback});
```

where the creation of the Directus client allows some kind of callback. So I can implement in the callback my own logic if the user is logged out due to a failed refresh.

**Reply by ched-dev** [reply]
OK that makes sense to me. You should update the original submission to outline the things you are asking for. Example: In the SDK you want to add a `onFailedRefresh` callback.

**Comment by br41nslug** [comment]
The feature you want, I think, is more generic shared error handling so you can catch any request where the API returns authentication errors (like token timed out etc, they can happen on any request not just refresh). Because the API is stateless the client does not know whether it is "logged in" or not, all it knows is whether it has a token or not (which may be expired) the only way to check "logged in status" is to do an actual call with the token and see if it does not error.

Currently you'll have to check for these errors manually and redirect to the login page if the API returns authentication errors.

**Reply by NilsBaumgartner1994** [reply]
Okay but how do i handle such errors at one specific place? Can you provide me a link to the doc?

**Reply by br41nslug** [reply]
Im just adding context and trying to get to the core of the request for implementation 😉 A centralized way of catching all SDK errors does not yet exist.

**Reply by NilsBaumgartner1994** [reply]
Oh no. What happend to the `transportHelper`?

**Reply by br41nslug** [reply]
the `transportHelper` hooked into `axios.transport` in the old SDK I think 🤔 and is not present in the current `fetch` based implementation. (there is a workaround however when using `ofetch` as fetcher for the SDK which does support `onResponseError` callbacks itself)

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
